File: 프로그래머스_거리두기확인하기.py
dx = [0,0,1,-1]
dy = [1,-1,0,0]

def bfs(place, i, j, distance):
    check = [[0]*5 for a in range(5)]
    q = []
    q.append([i,j,distance])
    
    while q :
        x, y, d= q.pop(0)
        check[x][y] = 1
        
        for a in range(4) :
            nx = x + dx[a]
            ny = y + dy[a]
            nd = d +1
            if 0<=nx<5 and 0<=ny<5 and not check[nx][ny]:
                if nd <= 2 and place[nx][ny] == 'P' :
                    return 0
                if place[nx][ny] != 'X' :
                    q.append([nx,ny,nd])
        
    
    return 1

def solution(places):
    answer = []
    
    for p in places :
        tmp = 1
        for i in range(5) :
            for j in range(5) :
                if p[i][j] == 'P' :
                    tmp2 = bfs(p, i, j, 0)
                    if not tmp2 :
                        tmp = 0
        answer.append(tmp)
    return answer

File: 프로그래머스_거리두기확인하기.cpp
#include <string>
#include <vector>
#include <queue>
#include<iostream>

using namespace std;

int dx[4] = {-1,0,1,0};
int dy[4] = {0,-1,0,1};

bool bfs(vector<string> p, int a, int b) {
    queue<pair<pair<int,int>,int>> q;
    bool ch[5][5] = {false,};
    
    q.push(make_pair(make_pair(a,b),0));
    ch[a][b] = 1;

    while(!q.empty()) {
        int x = q.front().first.first;
        int y = q.front().first.second;
        int d = q.front().second;
        q.pop();
        
        if(d >= 2) continue;
        
        for(int i=0; i<4; i++) {
            int nx = x+dx[i];
            int ny = y+dy[i];
            if(nx<0 || nx> 4 || ny<0 || ny >4) continue;
            
            if(!ch[nx][ny]){
                if(p[nx][ny] == 'P') {
                    return false;
                } else if(p[nx][ny] == 'O') {
                    q.push(make_pair(make_pair(nx,ny),d+1));
                    ch[nx][ny] = 1;
                }
            }
        }
    }
    
    
    return true;
}

int find_answer(vector<string> p) {
    for(int i=0; i<5; i++) {
        for(int j=0; j<5; j++) {
            if(p[i][j] == 'P') {
                if(bfs(p,i,j) == false) {
                    return 0;
                }
            }
        }
    }
    return 1;
}

vector<int> solution(vector<vector<string>> places) {
    vector<int> answer;
    
    bool flag = true;
    for(int i=0; i<places.size(); i++) {
        answer.push_back(find_answer(places[i]));
    }
    return answer;
}
